src/backend/columnar/columnar_customscan.c

```diff
@@ -712,6 +715,45 @@ ColumnarAttrNeeded(ScanState *ss)
 }
 
 
+/*
+ * ColumnarVarNeeded returns a list of Var objects for the ones that are
+ * needed during columnar custom scan.
+ * Throws an error if finds a Var referencing to an attribute not supported
+ * by ColumnarScan.
+ */
+static List *
+ColumnarVarNeeded(ScanState *ss)
+{
+	Plan *plan = ss->ps.plan;
+	int flags = PVC_RECURSE_AGGREGATES |
+				PVC_RECURSE_WINDOWFUNCS | PVC_RECURSE_PLACEHOLDERS;
+	List *varList = list_concat(pull_var_clause((Node *) plan->targetlist, flags),
+								pull_var_clause((Node *) plan->qual, flags));
+	ColumnarCheckVarListAttrNumSupported(varList);
+	return varList;
+}
+
+
+/*
+ * ColumnarCheckVarListAttrNumSupported throws an error if given varList
+ * references an attribute not supported by ColumnarScan.
+ */
+static void
+ColumnarCheckVarListAttrNumSupported(List *varList)
+{
+	Var *var = NULL;
+	foreach_ptr(var, varList)
+	{
+		if (var->varattno < 0)
```

Comment:
What about whole-row references (`attno == 0`)?

Reply:
Both callers (ColumnarAttrNeeded & the deparser function that we use for explain) are capable of handling attrNum=0

Reply:
But it seems like we need to do something, right? What if there's a table `foo(x int, y int)`. If we have a whole row reference, we want to show that we are projecting `x, y`. But if we have a whole row reference *and* a specific reference, like `x`, then I think it will end up printing something strange like `(x, y), x` or `foo, x` (I didn't look at the actual output).

In other words, if we have a whole row reference we want to turn that into all the individual Vars, and then dedupe, and then print them.

Reply:
>  then I think it will end up printing something strange like (x, y), x or foo, x

Currently, since `ColumnarVarNeeded` early returns for `attNum= 0`, we only show `foo.*` for that case, not `foo.*, x`.
(See the test that I added to `columnar_chunk_filtering.out`).

> In other words, if we have a whole row reference we want to turn that into all the individual Vars, and then dedupe, and then print them.

Right. Best way of doing that would be to expand `attNum=0` to individual `Var`s. But then we encounter with the same problem given here:
https://github.com/citusdata/citus/pull/4818#discussion_r693343001

For this reason, we should either print whole-row references in the form of `for.*`, or we should save `RelOptInfo` in somewhere and just build `Var`s from the output of `ColumnarAttrNeeded`, what do you think @jeff-davis ?